keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,795 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The authorization attributes will be extended to include field selectors and label selectors from
+Get, List, Watch, and Deletecollection.
+This will allow authorizers to use these selectors when making an authorization decision.
+
+## Motivation
+
+Security for per-node workloads could be improved by exposing field and label selectors to authorizers.
+Adding them as authorization attributes allows the development of new kinds of authorizers that
+leverage this information to provide security.
+In particular, it enables out-of-tree authorizers to experiment with ways to express restrictions based on field and label selectors.
+
+### Goals
+
+* Add field and label selectors to authorization attributes for Get, List, Watch, and DeleteCollection verbs.
+* Add field and label selectors to webhook authorization types.
+* Add field and label selectors to SelfSubjectAccessReview (SSAR) and SubjectAccessReview (SAR).
+  <<[UNRESOLVED deads2k, liggitt, aws-Micah: you guys volunteering to deliver this bit? ]>>
+* Update node authorizer to restrict on nodeName field selector.
+  <<[/UNRESOLVED]>>
+
+### Non-Goals
+
+* Create a generic in-tree authorizer that manages field or label selectors.
+
+## Proposal
+
+List, Watch, and DeleteCollection requests directly have field and label selector options.
+A single-item List or Watch request is remapped to the Get verb, so Get requests can have field and label selectors.
+
+### Authorization Attributes changes
+
+The authorization attributes have easy access to the query parameter field and label selectors.
+To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
+selector has no semantic meaning.
```

Comment:
what will kube-apiserver do if presented with a SubjectAccessReview that sets the selectors in combination with verbs other than list/watch/deletecollection?

Reply:
> what will kube-apiserver do if presented with a SubjectAccessReview that sets the selectors in combination with verbs other than list/watch/deletecollection?

I would like for webhooks to write code to handle field and label selection for any verb.  The kube-apiserver must only include field and label selectors in authorization attributes when they are honored.

This means that the SAR to the kube-apiserver must remove them from verbs where they aren't honored. But if the kube-apiserver grows the ability to honor them on say an update, then a webhook should also honor the restriction.  Ideally without modification.

Reply:
> if the kube-apiserver grows the ability to honor them on say an update, then a webhook should also honor the restriction. Ideally without modification.

That seems ideal, though to write code to honor this now, the semantics of what a selector on a write means probably have to be defined (must the selector match the incoming object, the existing object, either, or both?)

It's somewhat weird to define semantics for behavior which doesn't yet exist... the only prior art I can think of are `objectSelector` on admission. That takes the most conservative stance, if either old object or new object matches, the webhook is called:

> objectSelector is evaluated against both the oldObject and newObject, and is considered to match if either object matches the selector.

The most conservative stance on writes would be that both the old and new object must fall within the selector, I guess.

Reply:
> The kube-apiserver must only include field and label selectors in authorization attributes when they are honored.
>
> This means that the SAR to the kube-apiserver must remove them from verbs where they aren't honored.

So the SAR caller would be responsible for only populating the selector fields when they would be used, kube-apiserver wouldn't strip them out of presented SARs?

And SARs kube-apiserver generates internally would only include the selectors on verbs where they are used? I guess kube-apiserver will assume aggregated servers will support exactly what kube-apiserver does for list / watch selectors.

Reply:
> And SARs kube-apiserver generates internally would only include the selectors on verbs where they are used? I guess kube-apiserver will assume aggregated servers will support exactly what kube-apiserver does for list / watch selectors.

And the aggregated apiservers are explicitly told to authorize all requests themselves, so the aggregated APIs are well covered.



> So the SAR caller would be responsible for only populating the selector fields when they would be used, kube-apiserver wouldn't strip them out of presented SARs?

I rephrased on the push.

client -> kube-apiserver SAR endpoint: if the kube-apiserver would not honor the selectors, they are removed prior to calling the authorizer.

client -> webhook: the client is responsible for describing what they mean

kube-apiserver -> webhook: the kube-apiserver will send selectors for all verbs where the selector is honored.  I'll  add a section describing potential future semantics for field selector on Create, Update, and Delete.